File: python/solutions/sliding_window/longest_substring_no_repeat.py

class Solution:
    def lengthOfLongestSubstring(self, s: str):
        seen = set()
        left = 0
        best = 0

        for right, char in enumerate(s):
            while char in seen:
                seen.remove(s[left])
                left += 1
            seen.add(char)
            best = max(best, right - left + 1)

        return best

File: typescript/src/sliding-window/longest-substring-no-repeat.ts
export class LongestSubstringNoRepeat {
  lengthOfLongestSubstring(s: string): number {
    const seen = new Set<string>();
    let left = 0;
    let best = 0;

    for (let right = 0; right < s.length; right++) {
      while (seen.has(s[right])) {
        seen.delete(s[left]);
        left++;
      }
      seen.add(s[right]);
      best = Math.max(best, right - left + 1);
    }

    return best;
  }
}
